### PR TITLE
feat: add snippet index

### PR DESCRIPTION
**Pull request by busunkim96**
This PR adds `snippet_metadata.proto` and another samplegen utils class to store the snippets so that they can be looked up by library template code.

PRs to begin generating metadata and add the samples to the library docstrings will follow (I originally planned it for one PR, but the changeset was a bit too big).

### CODE REVIEW

**Review by software-dov — COMMENTED** [review]
Great job and nice tests.
I'm a little unclear about the dataflow here, please correct me if this is wrong.
1) A snippet is generated, either manually or by snippetgen
2) We send the snippet text into snippet_index.Snippet, which parses it and adds line number tags to particular sections.
3) (TODO) We update the generated documentation (and documentation generation code) to point to these particular regions as in-text assistance.

File: gapic/samplegen_utils/snippet_metadata.proto [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+  string full_name = 2;
+
+  // Indicates wether this method is synchronous or asynchronous.
+  // Some languages may support only one of the variants, in which case, this
+  // field will always have the same value (for that language).
+  bool async = 3;
+
+  // Parameters of this method in the same order as they appear on the method
+  // declaration. Must be empty if the method has no parameters.
+  repeated Parameter parameters = 4;
```
[the remark this answers is not in the document]
Comment (busunkim96) [reply gapic/samplegen_utils/snippet_metadata.proto:176]: @amanda-tarafa Which method parameters should be listed here?

File: tests/unit/samplegen/test_snippet_index.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+import pytest
+
+from gapic.samplegen_utils import snippet_metadata_pb2
+from gapic.samplegen_utils import snippet_index, types
+from common_types import DummyApiSchema, DummyService, DummyMethod
+
+
+@pytest.fixture
+def sample_str():
+    return"""# [START mollusc_classify_sync]
```
[the remark this answers is not in the document]
Comment (busunkim96) [reply tests/unit/samplegen/test_snippet_index.py:27]: Oops I didn't notice, thanks for catching this!

**Review by software-dov — APPROVED** [review]
LGTM, just one ongoing question and two test concatenations.

File: gapic/samplegen_utils/snippet_metadata.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply gapic/samplegen_utils/snippet_metadata.proto:176]: This are all the parameters for the **client** method the snippet is using, including all the call settings, etc.
Even if the snippet is not using some of the parameters of the client method, they should be included here.
If you have several client method overloads for the same RPC, the snippet for each overload should only list the parameters of the overload being used.
This is more to identify the client method than anything else.

**Review by amanda-tarafa — APPROVED** [review]
LGTM

File: gapic/samplegen_utils/snippet_metadata.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (software-dov) [reply gapic/samplegen_utils/snippet_metadata.proto:176]: Thanks for clearing that up.

**Review by software-dov — APPROVED** [review]

**Pull request merged**